Assorted mechanics and sound patches

kcat opened the pull request:
Apologies for the lack of focus in the patch set. Just went about fixing up some sound code issues and one thing led to the next...

kcat committed:
Avoid a hack to play the underwater sound properly

kcat committed:
Store the underwater sound to easily stop it

kcat committed:
Slightly randomize time between environment sounds

We should use the "Minimum Time Between Environmental Sounds" and
"Maximum Time Between Environmental Sounds" INI/fallback settings, but we don't
have them.

kcat committed:
Apply drowning damage based on the update duration

1 damage every 0.33 seconds is 3 damage a second. Applying it this way avoid
having to track another stat.

kcat committed:
Remove some unused functions

kcat committed:
Update the watched Ptr when changing it

kcat committed:
Cleanup some redundancy

kcat committed:
Avoid needlessly copying the MagicEffects

kcat committed:
Modify the current magicka and fatigue when the base changes

zinnschlag commented:
Looks okay. Except for the 2d array stuff. Can you please undo that? The 2d version is less efficient (uses a bit more space) and has worse maintainability (if a new element is added that is longer than the existing ones, you need to manually increase the 2nd dimension).

kcat commented:
Actually it'll be a bit inefficient either way since it's just going to be used to create a temporary `std::string` to pass to the WindowManager functions. For the dynamic stats, which can theoretically change every frame, it should be better to do something like
`static const std::string hbar("HBar");`
so the string construction only has to happen once. Would be nice if C++ had a copy-on-write string type like Qt does. For the attributes a `std::stringstream` would be more flexible, and they change so infrequently that the cost of constructing one as-needed is irrelevant.

kcat committed:
Avoid reconstructing strings for updating the dynamic stats

Attributes still do this, but they change infrequently enough that it doesn't
matter.

kcat committed:
Werewolves can't activate activators

zinnschlag commented:
I think you are missing my point. Efficiency is actually rather unimportant here. I can't see a reason even to bother with it. But static const char x[a][b] is inferior to static const char *x[a] in every single aspect, even if only slightly. I would even go so far and say there is no reason to use a 2-dimensional char array for a list of string literals ever.

I would prefer to not change something to something worse, even if it is only slightly worse, particularly if there was no reason to change anything in the first place. I am not convinced of your string solution either. More statics means more bookkeeping behind the scenes (since the compiler has to keep something like an extra flag for every static variable in function scope). And we are also losing a bit of flexibility, which might become a problem eventually, since this part of the code will be changed and extended a lot after 1.0.

Anyway, I'll go ahead and merge that. Don't want to spend even more time on something so trivial. Feels kinda frivolous.
